skorch/history.py

```diff
@@ -2,61 +2,58 @@
 
 
 # pylint: disable=invalid-name
-class _missingno:
-    def __init__(self, e):
-        self.e = e
-
-    def __repr__(self):
-        return 'missingno'
-
-
-def _incomplete_mapper(x):
-    for xs in x:
-        # pylint: disable=unidiomatic-typecheck
-        if type(xs) is _missingno:
-            return xs
-    return x
-
-
-# pylint: disable=missing-docstring
-def partial_index(l, idx):
-    needs_unrolling = (
-        isinstance(l, list) and len(l) > 0 and isinstance(l[0], list))
-    types = int, tuple, list, slice
-    needs_indirection = isinstance(l, list) and not isinstance(idx, types)
-
-    if needs_unrolling or needs_indirection:
-        return [partial_index(n, idx) for n in l]
-
-    # join results of multiple indices
-    if isinstance(idx, (tuple, list)):
-        zz = [partial_index(l, n) for n in idx]
-        if isinstance(l, list):
-            total_join = zip(*zz)
-            inner_join = list(map(_incomplete_mapper, total_join))
-        else:
-            total_join = tuple(zz)
-            inner_join = _incomplete_mapper(total_join)
-        return inner_join
-
-    try:
-        return l[idx]
-    except KeyError as e:
-        return _missingno(e)
-
-
-# pylint: disable=missing-docstring
-def filter_missing(x):
-    if isinstance(x, list):
-        children = [filter_missing(n) for n in x]
-        # pylint: disable=unidiomatic-typecheck
-        filtered = list(filter(lambda x: type(x) != _missingno, children))
-
-        if children and not filtered:
-            # pylint: disable=unidiomatic-typecheck
-            return next(filter(lambda x: type(x) == _missingno, children))
-        return filtered
-    return x
+class _none:
+    """Special placeholder since ``None`` is a valid value."""
+
+
+def _not_none(items):
+    """Whether the item is a placeholder or contains a placeholder."""
+    if not isinstance(items, (tuple, list)):
+        items = (items,)
+    return all(item is not _none for item in items)
+
+
+def _filter_none(items):
+    """Filter special placeholder value, preserves sequence type."""
+    type_ = list if isinstance(items, list) else tuple
+    return type_(filter(_not_none, items))
+
+
+def _getitem(item, i):
+    """Extract value or values from dicts.
+
+    Covers the case of a single key or multiple keys. If not found,
+    return placeholders instead.
+
+    """
+    if not isinstance(i, (tuple, list)):
+        return item.get(i, _none)
+    type_ = list if isinstance(item, list) else tuple
+    return type_(item.get(j, _none) for j in i)
+
+
+def _unpack_index(i):
+    """Unpack index and return exactly four elements.
+
+    If index is more shallow than 4, return None for trailing
+    dimensions.
+
+    """
+    # fill trailing indices with None
```

Comment:
Maybe assert that `len(i) <= 4`?

Reply:
yes, maybe we can check the length here, but I would rather raise a KeyError.

Reply:
Yeah, `KeyError` would be cleaner than an assertion I suppose. And definitely better than `ValueError: too many values to unpack` as it is right now.

Reply:
This now raises a `KeyError`.